src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,656 @@
+                                                                                                                         IMethodSymbol methodSymbol,
+                                                                                                                         CancellationToken cancellationToken)
+        {
+            var methodCallSites = new Dictionary<Document, List<TInvocationExpressionSyntax>>();
+            var progress = new StreamingProgressCollector();
+
+            await SymbolFinder.FindReferencesAsync(
+                methodSymbol, document.Document.Project.Solution, progress: progress,
+                documents: null, FindReferencesSearchOptions.Default, cancellationToken).ConfigureAwait(false);
+            var referencedSymbols = progress.GetReferencedSymbols();
+            var referencedLocations = referencedSymbols.SelectMany(referencedSymbol => referencedSymbol.Locations).Distinct().ToImmutableArray()
+                .OrderByDescending(reference => reference.Location.SourceSpan.Start);
+
+            var list = new List<TInvocationExpressionSyntax>();
+            methodCallSites.Add(document.Document, list);
+            foreach (var refLocation in referencedLocations)
+            {
+                if (!methodCallSites.TryGetValue(refLocation.Document, out var invocations))
+                {
+                    invocations = new List<TInvocationExpressionSyntax>();
+                    methodCallSites.Add(refLocation.Document, invocations);
+                }
+                invocations.Add((TInvocationExpressionSyntax)refLocation.Location.FindNode(cancellationToken).Parent!);
+            }
+
+            return methodCallSites.ToImmutableDictionary();
+        }
+
+        /// <summary>
+        /// Generates a method declaration containing a return expression of the highlighted expression
+        /// </summary>
+        private async Task IntroduceNewMethodAsync(SemanticDocument semanticDocument,
+                                                   TExpressionSyntax expression,
+                                                   IMethodSymbol methodSymbol,
+                                                   string newMethodIdentifier,
+                                                   SyntaxEditor editor,
+                                                   CancellationToken cancellationToken)
+        {
+            var document = semanticDocument.Document;
+            var syntaxFacts = semanticDocument.Document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var methodExpression = expression.FirstAncestorOrSelf<SyntaxNode>(node => IsMethodDeclaration(node), true)!;
```

Comment:
'methodExpression' isn't something that has meaning.  method declarations are not expressions.

Reply:
CHanged